twins/nats/pubsub.go

```diff
@@ -0,0 +1,138 @@
+//
+// Copyright (c) 2018
+// Mainflux
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package nats
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/twins"
+	"github.com/mainflux/mainflux/twins/paho"
+	"github.com/nats-io/go-nats"
+)
+
+const (
+	queue = "twins"
+	input = "channel.>"
+)
+
+type pubsub struct {
+	natsClient *nats.Conn
+	mqttClient paho.Mqtt
+	logger     log.Logger
+	twins      twins.TwinRepository
+	states     twins.StateRepository
+}
+
+// Subscribe to appropriate NATS topic
+func Subscribe(nc *nats.Conn, mc paho.Mqtt, tr twins.TwinRepository, sr twins.StateRepository, logger log.Logger) {
+	ps := pubsub{
+		natsClient: nc,
+		mqttClient: mc,
+		logger:     logger,
+		twins:      tr,
+		states:     sr,
+	}
+	ps.natsClient.QueueSubscribe(input, queue, ps.handleMsg)
+}
+
+func (ps pubsub) handleMsg(m *nats.Msg) {
+	var err error
+
+	var msg mainflux.Message
+	if err := proto.Unmarshal(m.Data, &msg); err != nil {
+		ps.logger.Warn(fmt.Sprintf("Unmarshalling failed: %s", err))
+		return
+	}
+	if msg.Channel == ps.mqttClient.Topic() {
+		return
+	}
+
+	b := []byte{}
+	id := ""
+	defer ps.mqttClient.Publish(&id, &err, "state/success", "state/failure", &b)
+
+	tw, err := ps.twins.RetrieveByThing(context.TODO(), msg.Publisher)
+	if err != nil {
+		ps.logger.Warn(fmt.Sprintf("Retrieving twin for %s failed: %s", msg.Publisher, err))
+		return
+	}
+
+	var recs []Record
```

Comment:
I've put it there because I've defined it right before the line where I use it. Seems clearer like this to me. I think that's how we do it generally. What do you think?

Reply:
Ah I see, I saw also some others var's declared at the fn begin so I was thinking that you defined all var's there and you missed this one :) All good.